### PR TITLE
FIX-#2830: read_csv with OmniSci backend compatibility with Pandas

### LINKED ISSUE #2830: read_csv with OmniSci backend compatibility with Pandas
Any call of read_csv with OmniSci backend with unsupported parameters should trigger pandas implementation.

### PR DESCRIPTION
**Pull request by amyskov**
## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2830  
- [x] tests added and passing

### CODE REVIEW

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 1]
```diff
@@ -288,3 +338,94 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+        ----------
+        read_csv_kwargs: ReadCsvKwargsType
+                Parameters of read_csv function.
+
+        Returns
+        -------
+        bool:
+            Whatever passed parameters are supported or not.
+        str:
+            Error message that should be raised if user explicitely set `engine="arrow"`.
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/io.py:361]: explicitly

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 2]
```diff
@@ -288,3 +338,94 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+
+        Parameters
+        ----------
+        read_csv_kwargs: ReadCsvKwargsType
+                Parameters of read_csv function.
+
+        Returns
+        -------
+        bool:
+            Whatever passed parameters are supported or not.
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/io.py:359]: Whatever -> Whether

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 3]
```diff
@@ -288,3 +338,94 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
 
         return usecols_md
+
+    @classmethod
+    def _read_csv_check_support(
+        cls,
+        read_csv_kwargs: ReadCsvKwargsType,
+    ) -> Tuple[bool, str]:
+        """
+        Check whatever or not passed parameters are supported by current modin.read_csv
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/io.py:348]: Check if passed parameters...

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 4]
```diff
@@ -288,3 +338,94 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+        names = read_csv_kwargs.get("names", None)
+        engine = read_csv_kwargs.get("engine", None)
+        skiprows = read_csv_kwargs.get("skiprows", None)
+        delimiter = read_csv_kwargs.get("delimiter", None)
+
+        if isinstance(filepath_or_buffer, str):
+            file_extension = os.path.splitext(filepath_or_buffer)[1]
+            if (
+                not cls.file_exists(filepath_or_buffer, only_locally=True)
+                or file_extension != ".csv"
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/io.py:374]: Why can't I load a `.txt` file? Does this restriction come from Pandas?

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 5]
```diff
@@ -288,3 +338,94 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+            ):
+                return False, cls.not_implemented_message
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/io.py:376]: If a file doesn't exist, we should provide a user with a better error message.

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 6]
```diff
@@ -288,3 +338,94 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+
+        read_csv_signature = inspect.signature(cls.read_csv)
+        read_csv_default_kwargs = {
+            k: v.default
+            for k, v in read_csv_signature.parameters.items()
+            if v.default is not inspect.Parameter.empty
+        }
+        for arg in cls.unsupported_args:
+            if read_csv_kwargs[arg] != read_csv_default_kwargs[arg]:
+                return False, cls.not_implemented_message
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/io.py:388]: Can we at least specify which parameter causes the trouble here and in other places?

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 7]
```diff
@@ -288,3 +338,94 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+            )
+            columns_number = len(empty_pandas_df.columns)
+            if header not in [None, 0, "infer"]:
+                return (
+                    False,
+                    "read_csv with 'arrow' engine and provided 'names' parameter supports only 0, None and 'infer' header values",
+                )
+            elif columns_number != len(names):
+                # pyarrow.lib.ArrowInvalid: CSV parse error: Expected 1 columns, got 6
+                return False, cls.not_implemented_message
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/io.py:415]: This is incorrect input provided by a user. `Not implemented` error message doesn't fit here. We probably should raise errors right in this method to use other types of error, e. g. `ValueError`

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 1 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Fixed.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 2 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: FIxed.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 3 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Corrected.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 4 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: `.txt` extension was just missed. Fixed in the last commit.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 5 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Ok, more specific comments were added.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 6 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Error messages were corrected.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 7 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: If we will raise `ValueError` in this method, it won't be catched by `try except` statements in the `modin.experimental.engines.omnisci_on_ray.io.read_csv` function and it will cause discrepancy with Pandas read_csv (Pandas handles such cases without exceptions). If specified error message is not enough in that case, we can create custom exception and add it to other catched exceptions [here](https://github.com/modin-project/modin/blob/master/modin/experimental/engines/omnisci_on_ray/io.py#L219)

Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/io.py]: I assumed Pandas would raise an error here. Why can't we follow Pandas's logic here (e.g. truncate names list)?

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 6 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/io.py]: What if a user explicitly chooses Arrow engine and provides a wrong file name? Will he get NotImplementedError?

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 5 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/io.py]: What if a user explicitly chooses Arrow engine and provides a wrong file name? Will he get NotImplementedError?

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 4 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Does Arrow really refuse to parse files with other extensions?

Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Arrow doesn't refuse most of the text formats, but for these formats both Pandas ([Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html)) and Arrow ([Arrow documentation](https://arrow.apache.org/docs/python/api/formats.html#json-files)) have specific functions (`read_json` or `read_fwf` for example), so, i think we don't need to handle it here.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 5 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Yes, he will. To avoid this confusion, i have added `ArrowEngineException` to raise it instead of `NotImplementedError` exception.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 4 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/io.py]: This is not about formats, this is about file names. Why can't I read CSV data from files named `data.tmp`, `my_csv_file`, etc.?

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 7 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: I think we can, but only after conversion from arrow table (after calling `from_arrow` function, we can add additional `NaN` column or combine columns), but it is better to separate this step to another PR.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 5 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Raising the same error for all issues is not very convenient. `FileNotFoundError` would be much more appropriate here and this would match Pandas' behavior. How about raising various errors right here instead of returning messages?

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 7 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Sure, these cases don't seem to be of high priority.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 4 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Ok, i see. You can read these files, but in that case (file extension is absent) we lose opportunity to check file type, that can be useful if user passes incorrect file type, for example
```
from pyarrow.csv import read_csv, ReadOptions, ConvertOptions
import pandas


test_filename = "modin/pandas/test/data/test_categories.json"
ans_pandas = pandas.read_csv(test_filename)
print(ans_pandas)
ans_arrow = read_csv(test_filename)

Output:
Empty DataFrame
Columns: [{"one":{"0":111, 1:222, 2:333}, two:{"0":"AAA", 1:"BBB", 2:"CCC"}}]
Index: []
Traceback (most recent call last):
  File "test.py", line 418, in <module>
    ans_arrow = read_csv(test_filename)
  File "pyarrow/_csv.pyx", line 714, in pyarrow._csv.read_csv
  File "pyarrow/error.pxi", line 122, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 84, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: CSV parse error: Empty CSV file or block: cannot infer number of columns
```
So it is needed to check file types. To sniff file type we can use third party packages, but i prefer to explicitly prohibit using of files without extensions like it is done in the last commit. What do you think?

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 5 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Ok, i think it is good idea. Exceptions were added in the `_read_csv_check_support` function.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 4 above)
Comment (ienkovich) [reply modin/experimental/engines/omnisci_on_ray/io.py]: From here I see that Pandas didn't discover the format and read it as if it were in CSV format. It didn't try to assume JSON format basing on the file extension. There is a `read_json` method to read JSON files.
There is a `compression` param which by default requires us to check for ‘.gz’, ‘.bz2’, ‘.zip’, and ‘.xz’ extensions to infer compression. I'm not sure if Arrow handles it for us. Otherwise, we shouldn't care about the file extension.

Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Ok, extension checks were removed and in order to track such cases (as was showed in the previous comment), `ArrowInvalid` exception was added to the catched exceptions list [here](https://github.com/amyskov/modin/blob/read_csv_omnisci_compatibility/modin/experimental/engines/omnisci_on_ray/io.py#L271).
Arrow infers `compression` automatically according to Arrow [read_csv docstring](https://github.com/apache/arrow/blob/master/python/pyarrow/_csv.pyx#L763-L766).

**Review by ienkovich — APPROVED** [review]
LGTM, Thanks!

**Review by ienkovich — APPROVED** [review]

**Comment by amyskov** [comment]
Hi @modin-project/modin-core please take a look at this PR, it mostly modifies `omnisci_on_ray` part, but it still affects IO tests owned by `modin-core` team.

File: modin/pandas/test/test_io.py [hunk 8]
```diff
@@ -899,13 +935,17 @@ def test_read_csv_default_to_pandas_url(self, request):
     @pytest.mark.parametrize("nrows", [21, 5, None])
     @pytest.mark.parametrize("skiprows", [4, 1, 500, None])
     def test_read_csv_newlines_in_quotes(self, nrows, skiprows):
+        if Backend.get() == "Omnisci" and skiprows in [4, 500] and nrows is None:
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:938]: `skiprows in [4, 500]` feels wrong - we'd have to always keep in sync `.parametrize()` above and this `.xfail()`.
```suggestion
        if Backend.get() == "Omnisci" and skiprows not [0, 1, None] and nrows is None:
```
I suggest listing supported values for `skiprows` instead

File: modin/pandas/test/test_io.py [hunk 9]
```diff
@@ -373,8 +381,24 @@ def test_read_csv_parsing_2(
...
+            and names is None
+            and skipfooter == 0
+            and nrows is None
+            and not true_values
+            and not false_values
+        )
+        if xfail_case_1:
             pytest.xfail("modin and pandas dataframes differs - issue #2446")
+        if xfail_case_2:
+            pytest.xfail("read_csv fails because of duplicated columns names")
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:401]: pls make a tracking issue and point to it here

File: modin/pandas/test/test_io.py [hunk 10]
```diff
@@ -428,6 +452,10 @@ def test_read_csv_squeeze(self, test_case):
         )
 
     def test_read_csv_mangle_dupe_cols(self):
+        if Backend.get() == "Omnisci":
+            pytest.xfail(
+                "processing of duplicated columns in OmniSci backend is not supported yet"
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:457]: submit a tracking issue and reference it here

File: modin/pandas/test/test_io.py [hunk 11]
```diff
@@ -483,6 +511,11 @@ def test_read_csv_datetime(
                 "The reason of tests fail in `cloud` mode is unknown for now - issue #2340"
             )
 
+        if Backend.get() == "Omnisci" and not isinstance(parse_dates, bool):
+            pytest.xfail(
+                "In some cases read_csv with `parse_dates` with OmniSci backend outputs incorrect result"
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:516]: submit a tracking issue and reference it here

File: modin/pandas/test/test_io.py [hunk 12]
```diff
@@ -855,7 +888,10 @@ def test_read_csv_s3(self):
     @pytest.mark.parametrize("names", [list("XYZ"), None])
     @pytest.mark.parametrize("skiprows", [1, 2, 3, 4, None])
     def test_read_csv_skiprows_names(self, names, skiprows):
-
+        if Backend.get() == "Omnisci" and names is None and skiprows in [1, None]:
+            pytest.xfail(
+                "processing of duplicated columns in OmniSci backend is not supported yet"
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:893]: submit a tracking issue and reference it here

File: modin/pandas/test/test_io.py [hunk 13]
```diff
@@ -953,6 +993,21 @@ def test_read_csv_names_neq_num_cols(self, request, kwargs):
             **kwargs,
         )
 
+    def test_read_csv_wrong_path(self):
+
+        raising_exceptions = list(io_ops_bad_exc).remove(FileNotFoundError)
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:998]: this doesn't work as you expect, `.remove()` always returns `None` while simultaneously modifying its class.
```suggestion
        raising_exceptions = list(set(io_ops_bad_exc) - set([FileNotFoundError]))
```
or
```suggestion
        raising_exceptions = [e for e in io_ops_bad_exc if e != FileNotFoundError]
```

File: modin/pandas/test/test_io.py [hunk 14]
```diff
@@ -953,6 +993,21 @@ def test_read_csv_names_neq_num_cols(self, request, kwargs):
...
+        eval_io(
+            fn_name="read_csv",
+            raising_exceptions=raising_exceptions,
+            # read_csv kwargs
+            filepath_or_buffer="/some/wrong/path.csv",
+        )
+
+    @pytest.mark.skipif(
+        Backend.get() == "Omnisci",
+        reason="to_csv is not implemented with OmniSci backend yet",
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:1009]: maybe submit a tracker and reference it here?

File: modin/pandas/test/test_io.py [hunk 15]
```diff
@@ -973,6 +1028,10 @@ def test_to_csv(self, request, header, mode):
             mode=mode,
         )
 
+    @pytest.mark.skipif(
+        Backend.get() == "Omnisci",
+        reason="to_csv is not implemented with OmniSci backend yet",
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:1033]: maybe submit a tracker and reference it here?

File: modin/pandas/test/test_io.py [hunk 16]
```diff
@@ -984,6 +1043,10 @@ def test_dataframe_to_csv(self, request):
             modin_obj=modin_df, pandas_obj=pandas_df, fn="to_csv", extension="csv"
         )
 
+    @pytest.mark.skipif(
+        Backend.get() == "Omnisci",
+        reason="to_csv is not implemented with OmniSci backend yet",
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:1048]: maybe submit a tracker and reference it here?

**Review by vnlitvinov — CHANGES_REQUESTED** [review]

**Comment by vnlitvinov** [comment]
@amyskov you now have a merge conflict, pls fix

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 17]
```diff
@@ -288,3 +343,130 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+                    f"read_csv with 'arrow' engine doesn't support {arg} parameter",
+                )
+        if delimiter is not None and read_csv_kwargs.get("delim_whitespace", False):
+            raise ValueError(
+                "Specified a delimiter with both sep and delim_whitespace=True; you can only specify one."
+            )
+        elif delimiter == "\n":
+            return (
+                False,
+                "read_csv with 'arrow' engine doesn't support delimiter = '\n' parameter",
```
Comment (dchigarev) [inline modin/experimental/engines/omnisci_on_ray/io.py:423]:
```suggestion
                "read_csv with 'arrow' engine doesn't support delimiter = '\\n' parameter",
```
if this message will be printed out somewhere `\n` will be interpreted as an actual new line character

File: modin/pandas/test/test_io.py [hunk 18]
```diff
@@ -16,13 +16,21 @@
...
 from modin.utils import to_pandas
 from modin.pandas.utils import from_arrow
 import pyarrow as pa
 import os
 import shutil
-import sqlalchemy as sa
+
+try:
+    import sqlalchemy as sa
+except ImportError:
```
Comment (dchigarev) [inline modin/pandas/test/test_io.py:28]: we should either add `sqlalchemy` in an OmniSci test environment file in this PR or create a tracker to do it in a separate PR and link it here

File: modin/pandas/test/test_io.py [hunk 19]
```diff
@@ -253,7 +261,7 @@ def test_read_csv_delimiters(
 
     # Column and Index Locations and Names tests
     @pytest.mark.skipif(
-        Engine.get() != "Python",
+        Engine.get() != "Python" and Backend.get() != "Omnisci",
```
Comment (dchigarev) [inline modin/pandas/test/test_io.py:264]:
```suggestion
        Engine.get() != "Python" or Backend.get() != "Omnisci",
```
shouldn't it be `or`?

File: modin/pandas/test/test_io.py (hunk 18 above)
Comment (vnlitvinov) [reply modin/pandas/test/test_io.py:28]: agreed

File: modin/pandas/test/test_io.py [hunk 20]
```diff
@@ -492,6 +522,16 @@ def test_read_csv_datetime(
                 reason="The reason of tests fail in `cloud` mode is unknown for now - issue #2340"
             )
 
+        if (
+            Backend.get() == "Omnisci"
+            and isinstance(parse_dates, list)
+            and ("col4" in parse_dates or 3 in parse_dates)
+        ):
+            # import pdb; pdb.set_trace()
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:530]: don't commit debug code

File: modin/pandas/test/test_io.py [hunk 21]
```diff
@@ -863,7 +903,10 @@ def test_read_csv_s3(self):
     @pytest.mark.parametrize("names", [list("XYZ"), None])
     @pytest.mark.parametrize("skiprows", [1, 2, 3, 4, None])
     def test_read_csv_skiprows_names(self, names, skiprows):
-
+        if Backend.get() == "Omnisci" and names is None and skiprows in [1, None]:
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:906]: did you forget to add `not` to `skiprows` check?

File: modin/pandas/test/test_io.py [hunk 22]
```diff
@@ -863,7 +903,10 @@ def test_read_csv_s3(self):
...
+            pytest.xfail(
+                "processing of duplicated columns in OmniSci backend is not supported yet - issue #3080"
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:908]: the message seems off - why it talks about duplicated columns and you check `skiprows`?

**Review by vnlitvinov — COMMENTED** [review]
Testing code looks better, I haven't looked at the implementation itself, though...
but OmniSci tests now fail in CI.

File: modin/pandas/test/test_io.py (hunk 10 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Fixed.

File: modin/pandas/test/test_io.py (hunk 11 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Fixed.

File: modin/pandas/test/test_io.py (hunk 12 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Fixed.

File: modin/pandas/test/test_io.py (hunk 14 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Fixed.

File: modin/pandas/test/test_io.py (hunk 15 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Fixed.

File: modin/pandas/test/test_io.py (hunk 16 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Fixed.

File: modin/pandas/test/test_io.py (hunk 18 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:28]: Not sure we should do these since `sqlalchemy` doesn't used by `Omnisci` backend. Do we need to add new package for tests compatibility only?

File: modin/pandas/test/test_io.py (hunk 20 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Fixed.

File: modin/pandas/test/test_io.py (hunk 21 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:905]: Not sure we should add `not` here, this test marked as `xfailed` because if one row will be skipped or rows won't be skipped at all, columns names will be inferred from the first row, that will contain duplicated values, that is not supported by  `Omnisci` backend yet. Corresponding comment is added.

File: modin/pandas/test/test_io.py (hunk 22 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:910]: Indeed this test is `xfailed` because of described reason (https://github.com/modin-project/modin/pull/2905#discussion_r637877703).

File: modin/pandas/test/test_io.py (hunk 18 above)
Comment (vnlitvinov) [reply modin/pandas/test/test_io.py:28]: I don't see anything wrong with adding test-only dependencies... for example, `pytest` isn't used by OmniSci, either 😉

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 23]
```diff
@@ -288,3 +343,126 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+        header = read_csv_kwargs.get("header", "infer")
+        names = read_csv_kwargs.get("names", None)
+        engine = read_csv_kwargs.get("engine", None)
+        skiprows = read_csv_kwargs.get("skiprows", None)
+        delimiter = read_csv_kwargs.get("delimiter", None)
+
+        if read_csv_kwargs.get("compression", "infer") != "infer":
+            return (
+                False,
+                "read_csv with 'arrow' engine doesn't support compression parameter, compression inferred"
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/io.py:377]:
```suggestion
                "read_csv with 'arrow' engine doesn't support compression parameter, compression must be inferred"
```

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 24]
```diff
@@ -288,3 +343,126 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+                        False,
+                        "read_csv with 'arrow' engine supports only local files",
+                    )
+                else:
+                    raise FileNotFoundError("No such file or directory")
+        elif not cls.pathlib_or_pypath(filepath_or_buffer):
+            if hasattr(filepath_or_buffer, "read"):
+                return (
+                    False,
+                    "read_csv with 'arrow' engine doesn't support file handles",
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/io.py:396]:
```suggestion
                    "read_csv with 'arrow' engine doesn't support file-like objects",
```

File: modin/pandas/test/test_io.py [hunk 25]
```diff
@@ -913,7 +958,7 @@ def test_read_csv_newlines_in_quotes(self, nrows, skiprows):
             filepath_or_buffer="modin/pandas/test/data/newlines.csv",
             nrows=nrows,
             skiprows=skiprows,
-            cast_to_str=True,
+            cast_to_str=True if Backend.get() != "Omnisci" else False,
```
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:961]:
```suggestion
            cast_to_str=Backend.get() != "Omnisci",
```
🤣

File: modin/pandas/test/test_io.py (hunk 18 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Ok, `sqlalchemy` was moved to the dependencies list.

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 26]
```diff
@@ -140,7 +205,11 @@ def read_csv(
         try:
             if eng in ["pandas", "c"]:
                 return cls._read(**mykwargs)
-
+            use_modin_impl, error_message = cls._read_csv_check_support(
+                mykwargs,
+            )
+            if not use_modin_impl:
+                raise ArrowEngineException(error_message)
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/io.py:212]: what is the point of _returning_ an error message to raise an exception? isn't it easier to just raise the exception in `_read_csv_check_support()` directly?

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 27]
```diff
@@ -288,3 +343,126 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+                    engine=None if engine == "arrow" else engine,
+                ),
+            )
+            columns_number = len(empty_pandas_df.columns)
+            if header not in [None, 0, "infer"]:
+                return (
+                    False,
+                    "read_csv with 'arrow' engine and provided 'names' parameter supports only 0, None and "
+                    "'infer' header values",
+                )
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/io.py:439]: move this check before `pandas.read_csv` call - you don't need the results of that call for this check, so it would be more efficient to skip the read if this check would trigger the return

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 26 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py:212]: This is done to maintain the common API of "check support" functions([case1](https://github.com/modin-project/modin/blob/master/modin/engines/base/io/text/csv_dispatcher.py#L176), [case2](https://github.com/modin-project/modin/blob/master/modin/engines/ray/generic/io.py#L52)). Does this make sense?

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 27 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Ok, fixed.

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 28]
```diff
@@ -11,19 +11,46 @@
...
+        Callable,
+        Dialect,
+        FilePathOrBuffer,
+        None,
+    ],
+]
+
 
+class ArrowEngineException(Exception):
+    """Exception should raised in the case if Arrow engine specific incompatibilities is found."""
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/io.py:50]:
```suggestion
    """Exception raised in case of Arrow engine-specific incompatibilities are found."""
```

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 26 above)
Comment (vnlitvinov) [reply modin/experimental/engines/omnisci_on_ray/io.py:212]: Okay, let's make a "cleanup issue" to not clutter this PR.

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 29]
```diff
@@ -288,3 +343,127 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+        header = read_csv_kwargs.get("header", "infer")
+        names = read_csv_kwargs.get("names", None)
+        engine = read_csv_kwargs.get("engine", None)
+        skiprows = read_csv_kwargs.get("skiprows", None)
+        delimiter = read_csv_kwargs.get("delimiter", None)
+
+        if read_csv_kwargs.get("compression", "infer") != "infer":
+            return (
+                False,
+                "read_csv with 'arrow' engine doesn't support compression parameter, compression must be inferred"
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/io.py:377]:
```suggestion
                "read_csv with 'arrow' engine doesn't support explicit compression parameter, compression must be inferred"
```

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 30]
```diff
@@ -288,3 +343,127 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+                " automatically (supported compression types are gzip and bz2)",
+            )
+
+        if isinstance(filepath_or_buffer, str):
+            if not os.path.exists(filepath_or_buffer):
+                from pandas.io.common import is_url
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/io.py:383]: I suggest doing imports at the top of file unless there are really good reasons not to.

File: modin/experimental/engines/omnisci_on_ray/io.py [hunk 31]
```diff
@@ -288,3 +343,127 @@ def _prepare_pyarrow_usecols(cls, read_csv_kwargs):
...
+                raise ValueError(
+                    f"Invalid file path or buffer object type: {type(filepath_or_buffer)}"
+                )
+
+        read_csv_signature = inspect.signature(cls.read_csv)
+        read_csv_default_kwargs = {
+            k: v.default
+            for k, v in read_csv_signature.parameters.items()
+            if v.default is not inspect.Parameter.empty
+        }
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_ray/io.py:408]: I suggest making this a class-level attribute computed after `read_csv` declaration, as this won't ever change - no need to do the same (potentially intensive) computation each time a user wants to read a file; you can also use `unsupported_args` to filter supported ones (effectively leaving only the unsupported subset of kwargs which have default values), this would lead to something like

```python
class OmniSciOnRay(...):
    # some class attrs here
    def read_csv(...):
        # read_csv body
    # somewhere in the class body
    read_csv_defaults = {k: v.default for k, v in inspect.signature(read_csv).parameters.items() if v.default is not inspect.Parameter.empty and k in unsupported_args}
    def _read_csv_check_support(...):
        # body
        for arg, def_value in cls.read_csv_defaults.items():
            if read_csv_kwargs[arg] != def_value:
                return False, "bad arg"
```

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 26 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py:212]: Ok, issue was created https://github.com/modin-project/modin/issues/3166.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 30 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Import was moved to the top of the file.

File: modin/experimental/engines/omnisci_on_ray/io.py (hunk 31 above)
Comment (amyskov) [reply modin/experimental/engines/omnisci_on_ray/io.py]: Sounds reasonable, suggestion was implemented (`read_csv_defaults` dict comprehension was replaced with loop to avoid 
comprehension scope problem with `unsupported_args`).

**Review by vnlitvinov — APPROVED** [review]
LGTM

**Review by devin-petersohn — APPROVED** [review]
